### DISCUSSION TITLE
Binaryen as WASM module itself?

### DISCUSSION THREAD

**Discussion by marad** [WebAssembly/binaryen#5389] [Ideas]
I'd love to use binaryen from Java. I've even started preparing a wrapper, which is kinda painful since I've had to generate Windows DLL by tweaking cmake.

Then it hit me - it would all be solved by having binaryen be a wasm module as well, since there are solutions to run WASM from Java. 

Is something like this possible?

**Comment by tlively** [comment]
Yes, and it's even supported today. If you compile the `binaryen_wasm` target, you'll get a Wasm module that exports Binaryen's C API as well as a JS wrapper providing an idiomatic JS API on top of it. You could always discard the JS and reuse the Wasm module directly, though.

**Reply by marad** [reply]
I was excited at first, but it doesn't really work this way. This target only generates single JS file which seems to have base64 encoded wasm module right in the JS file ;( 

Of course it's possible to extract the bytes with some JS-fu, but it would be better to simply have target to generate basic `binaryen.wasm` as well.

(or it's possible that I just don't know how to build it ;) )

**Reply by tlively** [reply]
Oh, I forgot that we built with `-sSINGLE_FILE`. Removing the `add_link_flag("-sSINGLE_FILE")` line from CMakeLists.txt should fix that and give you a separate .wasm file.

**Reply by marad** [reply]
Awesome! I'll check that out later today! Thanks :)

**Reply by marad** [reply]
Ok, this worked like a charm. I'll leave the instructions here in case anybody wanted to perform this stunt again (or even me in the future) ;) 

1) Clone binaryen repo and do `git submodule init` && `git submodule update` as instructed in the readme
2) Edit `CMakeLists.txt` and remove/comment out the `add_link_flag("-sSINGLE_FILE")` line. (This makes it actually build the wasm file instead of embedding it into JS file)
3) Run `emcmake cmake .` I used docker for this:

```bash
docker run \
	--rm \
	-v $(pwd):/src \
	-u $(id -u):$(id -g) \
	emscripten/emsdk \
	emcmake cmake .
```

4) And make the thing with `emmake make -j10 binaryen_wasm`. I also used docker for this:

```bash
docker run \
	--rm \
	-v $(pwd):/src \
	-u $(id -u):$(id -g) \
	emscripten/emsdk \
	emmake make -j10 binaryen_wasm
```

Of course adjust the number of threads (the `-j10` flag) accordingly to your CPU.

As a result the `bin` folder is going to have `binaryen_wasm.wasm` file along with the JS glue code.

Again - thanks @tlively for supporting me on this.

**Reply by marad** [reply]
@tlively Following up on this one. I wanted to use the new module, but it seems to want to have ~80 functions imported, but the function names are 'a', 'b', and so on. I suspect there is some name obfuscation in play. 

What are those functions that the module requires and how to maybe disable obfuscation so at least the names are comprehensible?

**Reply by tlively** [reply]
@kripken, do you know what the best set of flags to use here is?

**Reply by kripken** [reply]
`--profiling` or `--profiling-funcs` will disable name minification.

**Reply by marad** [reply]
Nice, I'll definately check it out. I suspect the functions are IO functions. I'll probably also try compiling with in-memory FS to have less of them to integrate.

**Reply by stephanos** [reply]
@marad did you ever get all of that to work? if you did and could share any insights into what remaining steps you had to take to make it work end-to-end, I'd appreciate it!

**Reply by marad** [reply]
@stephanos I did not work on this much further unfortunately, but I would like to try again in some time. What I'm planning on doing is on top of adding `--profiling` and `--profiling-funcs` I remember I saw one option to use in-memory FS. I suppose (haven't checked that) this could result in the module requiring a memory block to work with instead of importing all the FS functions.

Then I'd take a look at what other imports it needs and improvise ;)

If you try this, let me know how it went. Maybe it would be sufficient to use it this way.

**Comment by marad** [comment]
I've got some time today to fiddle with this and I must say that binaryen is trying hard not to get compiled to basic WASM module ;)

I've made quite a few variations of the `CMakeLists.txt` but the most imortant ones seem to be:

* removing the "-sSINGLE_FILE"
* setting `set(CMAKE_EXECUTABLE_SUFFIX ".wasm")` before `add_executable(binaryen_wasm...`
* adding `target_link_libraries(binaryen_wasm "-sSIDE_MODULE=2")`

This gets me a WASM file without any JS file, and now I can see the names of the functions. 

It's not clear to me if that is a pure WASI since there are functions like `malloc` but also `invoke_viiiiiiii` which seem like those that were generated glue-code from JS.

```
  ...
  (import "env" "__cxa_atexit" (func $__cxa_atexit (type 6)))
  (import "env" "__cxa_begin_catch" (func $__cxa_begin_catch (type 2)))
  (import "env" "__cxa_end_catch" (func $__cxa_end_catch (type 16)))
  (import "env" "_ZSt9terminatev" (func $std::terminate__ (type 16)))
  (import "env" "strncpy" (func $strncpy (type 6)))
  (import "env" "malloc" (func $malloc (type 2)))
  (import "env" "pthread_self" (func $pthread_self (type 5)))
  (import "env" "free" (func $free (type 1)))
  (import "env" "posix_memalign" (func $posix_memalign (type 6)))
  (import "env" "__cxa_allocate_exception" (func $__cxa_allocate_exception (type 2)))
  (import "env" "invoke_viiiiiiii" (func (;108;) (type 67)))
  ...
```

@stephanos FYI ☝️

**Reply by marad** [reply]
Also for future reference: 

- [emcc options](https://emscripten.org/docs/tools_reference/emcc.html)
- [emcc settings](https://github.com/emscripten-core/emscripten/blob/main/src/settings.js)

**Reply by kripken** [reply]
To avoid the invokes, enable wasm exception handling instead of emscripten:

https://emscripten.org/docs/porting/exceptions.html#webassembly-exception-handling-based-support

Not all VMs support it yet, but it is the only way to use C++ exceptions without emscripten's `invoke_*` imports.

**Reply by stephanos** [reply]
@marad Thanks for the update! I won't get to try this for another couple of weeks, but I'm excited to now!

**Comment by artyompetrov** [comment]
Hello! did you make it?

**Reply by marad** [reply]
Nope, I was hoping for some more determined individual from the internet to pick that up from where I left it :D
